### PR TITLE
Trace and fix server sharing of Moped::Session connections between threads

### LINKED ISSUE #1964: Server threads share Moped::Session connections
The server runs in a clustered puma configuration with two worker processes, each running a number of threads, including Celluloid threads, EventMachine threads, as well as custom threads. The `Moped::Session` class is not thread-safe: https://groups.google.com/forum/#!topic/mongoid/0pIRfVjTL_Y

> With Moped, you are "thread safe" as long as you only have 1 moped session instance on each thread, and don't share a Moped::Session across threads.

> With Mongoid, we automatically manage that session-per-thread requirement for Moped for you, so you would not need to worry about it in this case. If a new thread is spawned, Mongoid will create a new session on the thread.

Mongoid uses `Thread.current["[mongoid]:sessions"]` to keep track of a per-thread `Moped::Session`, ensuring that sessions are not shared across threads, and each thread has its own session. However, storing the result of something like `@db_session = ContainerLog.collection.session` bypasses this protection, meaning that the `@db_session` is now shared between any threads that access it:

```ruby
module Agent
  class MessageHandler
    attr_reader :db_session

    def initialize(...)
      ...
      @db_session = ContainerLog.collection.session.with(
        ...
      )
    end

    def run
      Thread.new {
        .... # something that calls on_node_stat(...)
      }
    end

    def on_node_stat(...)
      ....
      db_session[:host_node_stats].insert(stat)
    end
  end
end
```

Using the `Moped::Session` thread tracing implemented as part of #1965 reveals two such cases for a quick test of Kontena 1.1.4 with a single node connected.

The `Moped::Session[8] = ...` part shows which thread created the session and where. The following  Moped::Session[8] ! ...` part shows the conflicting access by a different thread.

## `MongoPubsub`
```
Moped::Session[8] = process=7 thread=#<Thread:0x0055d08cf862d8>:
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `new'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `create_session'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:43:in `default'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions.rb:37:in `default'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid.rb:65:in `default_session'
	/app/app/initializers/00_mongoid.rb:6:in `<top (required)>'
	/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	/app/app/boot.rb:22:in `block in require_glob'
	/app/app/boot.rb:21:in `each'
	/app/app/boot.rb:21:in `require_glob'
	/app/app/boot.rb:26:in `<top (required)>'
	config/puma.rb:8:in `require_relative'
	config/puma.rb:8:in `block in _load_from'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `block in run_hooks'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `each'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `run_hooks'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:237:in `worker'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `block (2 levels) in spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `fork'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `block in spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:116:in `times'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:116:in `spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:426:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/launcher.rb:172:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cli.rb:74:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/bin/puma:10:in `<top (required)>'
	/usr/bin/puma:22:in `load'
	/usr/bin/puma:22:in `<main>'
Moped::Session[8] ! process=7 thread=#<Celluloid::Thread:0x0055d090683b78>:
	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/session.rb:61:in `collection_names'
	/app/app/services/mongo_pubsub.rb:170:in `ensure_collection!'
	/app/app/services/mongo_pubsub.rb:142:in `tail!'
	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:26:in `public_send'
	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:26:in `dispatch'
	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:122:in `dispatch'
	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.16.0/lib/celluloid/cell.rb:60:in `block in invoke'
	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.16.0/lib/celluloid/cell.rb:71:in `block in task'
	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:357:in `block in task'
	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.16.0/lib/celluloid/tasks.rb:57:in `block in initialize'
	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.16.0/lib/celluloid/tasks/task_fiber.rb:15:in `block in create'
```

## Kontena 1.1.x `Agent::MessageHandler`

This does not affect the current master RPC handlers, as those are Celluloid actors, which create their own session in their `initializer`?

```
Moped::Session[8] = process=11 thread=#<Thread:0x0055d08cf862d8>:
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `new'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `create_session'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:43:in `default'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions.rb:37:in `default'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid.rb:65:in `default_session'
	/app/app/initializers/00_mongoid.rb:6:in `<top (required)>'
	/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	/app/app/boot.rb:22:in `block in require_glob'
	/app/app/boot.rb:21:in `each'
	/app/app/boot.rb:21:in `require_glob'
	/app/app/boot.rb:26:in `<top (required)>'
	config/puma.rb:8:in `require_relative'
	config/puma.rb:8:in `block in _load_from'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `block in run_hooks'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `each'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `run_hooks'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:237:in `worker'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `block (2 levels) in spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `fork'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `block in spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:116:in `times'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:116:in `spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:426:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/launcher.rb:172:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cli.rb:74:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/bin/puma:10:in `<top (required)>'
	/usr/bin/puma:22:in `load'
	/usr/bin/puma:22:in `<main>'
Moped::Session[8] ! process=11 thread=#<Thread:0x0055d08fb9b0d0>:
	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/session.rb:50:in `[]'
	/app/app/services/agent/message_handler.rb:250:in `grid_container'
	/app/app/services/agent/message_handler.rb:100:in `on_container_event'
	/app/app/services/agent/message_handler.rb:64:in `handle_message'
	/app/app/services/agent/message_handler.rb:34:in `block (2 levels) in run'
	/app/app/services/agent/message_handler.rb:31:in `loop'
	/app/app/services/agent/message_handler.rb:31:in `block in run'
```

```
Moped::Session[8] ! process=11 thread=#<Thread:0x0055d08fb9b0d0>:
	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/session.rb:50:in `[]'
	/app/app/services/agent/message_handler.rb:250:in `grid_container'
	/app/app/services/agent/message_handler.rb:239:in `cached_container'
	/app/app/services/agent/message_handler.rb:116:in `on_container_log'
	/app/app/services/agent/message_handler.rb:66:in `handle_message'
	/app/app/services/agent/message_handler.rb:34:in `block (2 levels) in run'
	/app/app/services/agent/message_handler.rb:31:in `loop'
	/app/app/services/agent/message_handler.rb:31:in `block in run'
Moped::Session[8] = process=11 thread=#<Thread:0x0055d08cf862d8>:
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `new'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `create_session'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:43:in `default'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions.rb:37:in `default'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid.rb:65:in `default_session'
	/app/app/initializers/00_mongoid.rb:6:in `<top (required)>'
	/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	/app/app/boot.rb:22:in `block in require_glob'
	/app/app/boot.rb:21:in `each'
	/app/app/boot.rb:21:in `require_glob'
	/app/app/boot.rb:26:in `<top (required)>'
	config/puma.rb:8:in `require_relative'
	config/puma.rb:8:in `block in _load_from'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `block in run_hooks'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `each'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `run_hooks'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:237:in `worker'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `block (2 levels) in spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `fork'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `block in spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:116:in `times'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:116:in `spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:426:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/launcher.rb:172:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cli.rb:74:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/bin/puma:10:in `<top (required)>'
	/usr/bin/puma:22:in `load'
	/usr/bin/puma:22:in `<main>'
Moped::Session[8] ! process=11 thread=#<Thread:0x0055d08fb9b0d0>:
	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/session.rb:50:in `[]'
	/app/app/services/agent/message_handler.rb:140:in `flush_logs'
	/app/app/services/agent/message_handler.rb:134:in `on_container_log'
	/app/app/services/agent/message_handler.rb:66:in `handle_message'
	/app/app/services/agent/message_handler.rb:34:in `block (2 levels) in run'
	/app/app/services/agent/message_handler.rb:31:in `loop'
	/app/app/services/agent/message_handler.rb:31:in `block in run'
```

```
Moped::Session[8] = process=11 thread=#<Thread:0x0055d08cf862d8>:
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `new'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `create_session'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:43:in `default'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions.rb:37:in `default'
	/usr/lib/ruby/gems/2.3.0/gems/mongoid-4.0.2/lib/mongoid.rb:65:in `default_session'
	/app/app/initializers/00_mongoid.rb:6:in `<top (required)>'
	/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	/app/app/boot.rb:22:in `block in require_glob'
	/app/app/boot.rb:21:in `each'
	/app/app/boot.rb:21:in `require_glob'
	/app/app/boot.rb:26:in `<top (required)>'
	config/puma.rb:8:in `require_relative'
	config/puma.rb:8:in `block in _load_from'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `block in run_hooks'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `each'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:271:in `run_hooks'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:237:in `worker'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `block (2 levels) in spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `fork'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:120:in `block in spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:116:in `times'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:116:in `spawn_workers'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cluster.rb:426:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/launcher.rb:172:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/cli.rb:74:in `run'
	/usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/bin/puma:10:in `<top (required)>'
	/usr/bin/puma:22:in `load'
	/usr/bin/puma:22:in `<main>'
Moped::Session[8] ! process=11 thread=#<Thread:0x0055d08fb9b0d0>:
	/usr/lib/ruby/gems/2.3.0/gems/moped-2.0.7/lib/moped/session.rb:50:in `[]'
	/app/app/services/agent/message_handler.rb:205:in `on_node_stat'
	/app/app/services/agent/message_handler.rb:60:in `handle_message'
	/app/app/services/agent/message_handler.rb:34:in `block (2 levels) in run'
	/app/app/services/agent/message_handler.rb:31:in `loop'
	/app/app/services/agent/message_handler.rb:31:in `block in run'
```

### PR DESCRIPTION
**Pull request by SpComb**
Fixes #1964 for currently known cases by moving the `ModelClass.collection` calls into the same thread context where the resulting `Moped::Session` is used, using Mongoid's thread-local sessions to avoid sharing the same `Moped::Session` objects across threads.

Backport of the fixes for 1.1.5 in #1966

## Tracing
When running in DEBUG mode, the server loads a `MopedSessionTracer` early during puma worker bootup, which prints out stack traces of any shared `Moped::Session`s. This is also used in fatal aborting mode for the rspecs:

```
THREAD CONFLICT ON Moped::Session[8] @ process=10387 thread=#<Celluloid::Thread:0x00000002e69c70>:
	/home/kontena/kontena/kontena/server/lib/thread_tracer.rb:65:in `check!'
	/home/kontena/kontena/kontena/server/lib/moped_session_tracer.rb:13:in `current_database'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/session.rb:61:in `collection_names'
	/home/kontena/kontena/kontena/server/app/services/mongo_pubsub.rb:172:in `ensure_collection!'
	/home/kontena/kontena/kontena/server/app/services/mongo_pubsub.rb:144:in `tail!'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
OWNING THREAD process=10387 thread=#<Thread:0x00000001ad21a8>:
	/home/kontena/kontena/kontena/server/lib/moped_session_tracer.rb:8:in `new'
	/home/kontena/kontena/kontena/server/lib/moped_session_tracer.rb:8:in `initialize'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `new'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:66:in `create_session'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions/factory.rb:43:in `default'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid/sessions.rb:37:in `default'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongoid-4.0.2/lib/mongoid.rb:65:in `default_session'
	/home/kontena/kontena/kontena/server/app/initializers/00_mongoid.rb:6:in `<top (required)>'
	/home/kontena/kontena/kontena/server/app/boot.rb:22:in `require'
	/home/kontena/kontena/kontena/server/app/boot.rb:22:in `block in require_glob'
	/home/kontena/kontena/kontena/server/app/boot.rb:21:in `each'
	/home/kontena/kontena/kontena/server/app/boot.rb:21:in `require_glob'
	/home/kontena/kontena/kontena/server/app/boot.rb:26:in `<top (required)>'
	/home/kontena/kontena/kontena/server/spec/spec_helper.rb:32:in `require_relative'
	/home/kontena/kontena/kontena/server/spec/spec_helper.rb:32:in `<top (required)>'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1072:in `require'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1072:in `block in requires='
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1072:in `each'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1072:in `requires='
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration_options.rb:103:in `block in process_options_into'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration_options.rb:102:in `each'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration_options.rb:102:in `process_options_into'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/configuration_options.rb:22:in `configure'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:95:in `setup'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:84:in `run'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:69:in `run'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:37:in `invoke'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.1.7/exe/rspec:4:in `<top (required)>'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/bin/rspec:23:in `load'
	/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/bin/rspec:23:in `<main>'
```

## TODO

* Figure out if there are any other shared `Moped::Session`s that need fixing...
* Should we always load the `MopedSessionTracer`, also in production mode, not just DEBUG?

### CODE REVIEW

**Comment by SpComb** [comment]
Rebased this for 1.2 master, separate PR For 1.1.5 including the `Agent::MessageHandler` fix: #1966

**Comment by SpComb** [comment]
Somehow the changes here are triggering various mongodb errors in the server rspecs, including ones similar to those seen in https://github.com/kontena/kontena/issues/1948#issuecomment-286127161

* https://travis-ci.org/kontena/kontena/jobs/210622646#L658
```
  An error occurred in an after hook    NoMethodError: undefined method `[]' for nil:NilClass    occurred at /home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/moped-2.0.7/lib/moped/protocol/reply.rb:71:in `command_failure?'
```


* https://travis-ci.org/kontena/kontena/jobs/210595262#L710
```
Moped::Errors::Operationfailure ... failed with error nil
```

```
`KeyError: key not found: "a"`
```

Still need to figure out what's going on exactly. I might suspect that the changes to the `MongoPubsub` to always create a new `Moped::Session` for each rspec example may cause this?

**Comment by SpComb** [comment]
Not seeing any threading conflicts when running the e2e tests with `DEBUG=true`.

**Comment by SpComb** [comment]
The rspec failures on Moped errors above are likely caused by #1986... the thread fixes in this PR may or may not be related to those timeout/exception-safety issues, there may be some interactions that causes those to happen more or less often.

**Comment by SpComb** [comment]
@jakolehm PTAL now that #1987 fixes the related timeout issues

File: server/lib/thread_tracer.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    def to_s
+      "process=#{@process} thread=#{@thread}:\n#{@stack.map{|x|"\t#{x}"}.join("\n")}"
+    end
+  end
+
+  # Abort on errors
+  # Intended for test and development
+  def self.fatal!
+    @fatal = true
+  end
```
Comment (jakolehm) [inline server/lib/thread_tracer.rb]: coding style: missing new line

**Pull request merged**